src/components/Table/utils/TableState.ts

```diff
@@ -342,3 +408,58 @@ function flattenRows(rows: GridDataRow<any>[]): GridDataRow<any>[] {
   const childRows = rows.flatMap((r) => (r.children ? flattenRows(r.children) : []));
   return [...rows, ...childRows];
 }
+
+// Exported for testing purposes
+export function deriveSortState(
+  currentSortState: SortState,
+  clickedKey: string,
+  initialSortState: SortState | undefined,
+): SortState | undefined {
+  // If the current sort state is not defined then sort ASC on the clicked key.
+  if (!currentSortState.current) {
+    return { ...initialSortState, current: { columnId: clickedKey, direction: ASC } };
+  }
+
+  const {
+    current: { columnId: currentKey, direction: currentDirection },
+  } = currentSortState;
+
+  // If clicking a new column, then sort ASC on the clicked key
+  if (clickedKey !== currentKey) {
+    return { ...initialSortState, current: { columnId: clickedKey, direction: ASC } };
+  }
+
+  // If there is an `initialSortState` and we're clicking on that same key, then flip the sort.
+  // Handles cases where the initial sort is DESC so that we can allow for DESC to ASC sorting.
+  if (initialSortState && initialSortState.current?.columnId === clickedKey) {
+    return {
+      ...initialSortState,
+      current: { columnId: clickedKey, direction: (currentDirection as any as string) === ASC ? DESC : ASC },
+    };
+  }
+
+  // Otherwise when clicking the current column, toggle through sort states
+  if ((currentDirection as any as string) === ASC) {
+    // if ASC -> go to desc
+    return { ...initialSortState, current: { columnId: clickedKey, direction: DESC } };
+  }
+
+  // Else, direction is already DESC, so revert to original sort value.
+  return initialSortState;
+}
+
+type ColumnSort = {
+  columnId: string;
+  direction: Direction;
+};
+
+export type SortState = {
+  current?: ColumnSort;
+  persistent?: ColumnSort;
+};
+
+export type SortOn = "client" | "server" | undefined;
+
+function isSortState(maybeSortState: SortState | {}): maybeSortState is SortState {
```

Comment:
Question: Do we still need this now that `TableStart.sort` is always `SortState`?

Reply:
Oh, shoot. I thought I deleted it. I think I removed its uses. Thanks for pointing out.